### PR TITLE
Update rollup to v3

### PR DESCRIPTION
**Pull request by ymqy**
## Summary

For the rollup upgrade, I may not be very confident that this can be done, for the configuration changes are only for the ci all through, but there are still some problems need to line up, why some build files deleted does not affect the ci results, ci execution results and local execution results are not consistent, see the following comments

Closes https://github.com/facebook/react/pull/26078

## How did you test this change?

ci green

### CODE REVIEW

**Comment by ymqy** [comment]
Before upgrading Rollup, Jest must first be upgraded to version v27 or above, otherwise the test cases will not pass because Jest v26 does not support the 'node:' protocol, please see https://github.com/facebook/jest/issues/11339.

**Comment by ymqy** [comment]
the bundle size has increased, there may be a problem there and it needs to be investigated.the bundle size has increased, there may be a problem there and it needs to be investigated.

**Review by kassens — CHANGES_REQUESTED** [review]
Hey @ymqy thanks for the PRs to bring some overdue updates! This PR is a huge and a difficult to review. I think you've already started separating some smallers parts of these changes into new PRs. I think this'll help a lot.

For example, I see jest related changes in this rollup PR. I think if we can separate them, it'll be a lot easier to review and merge! Thanks for the help!

**Comment by ymqy** [comment]
> Hey @ymqy thanks for the PRs to bring some overdue updates! This PR is a huge and a difficult to review. I think you've already started separating some smallers parts of these changes into new PRs. I think this'll help a lot.
> 
> For example, I see jest related changes in this rollup PR. I think if we can separate them, it'll be a lot easier to review and merge! Thanks for the help!

Thanks for the suggestion, I am preparing to split into smaller separate pr's to make review and merge easier

**Comment by ymqy** [comment]
@kassens Before upgrading rollup, I need to upgrade jest. I originally only wanted to upgrade rollup, but during the upgrade process, I encountered that jest v26 does not support the 'node:' protocol, so I upgraded jest first. I split the jest upgrade into multiple smaller PRs. Now the CI has passed and is ready for review. After merging, I am planning to submit a PR for jest configuration optimization. https://github.com/facebook/react/pull/26088

**Comment by ymqy** [comment]
please review https://github.com/facebook/react/pull/26088

**Comment by ymqy** [comment]
> @ymqy Would be awesome if you could rebase this PR. We still want to land this but updating build tooling needs extra care to not accidentally break releases.

Of course!

File: packages/react-client/src/forks/ReactFlightClientHostConfig.dom-node.js [hunk 1]
```diff
@@ -9,4 +9,4 @@
 
 export * from 'react-client/src/ReactFlightClientHostConfigBrowser';
 export * from 'react-client/src/ReactFlightClientHostConfigStream';
-export * from 'react-server-dom-webpack/src/ReactFlightClientWebpackBundlerConfig';
+export * from 'react-server-dom-webpack/src/ReactFlightClientWebpackBundlerConfig.js';
```
Comment (ymqy) [inline packages/react-client/src/forks/ReactFlightClientHostConfig.dom-node.js:12]: I added the '.js' file extension to the import path to fix the ENOENT error. But it is not clear why the '.js' file extension is not added automatically

image

File: packages/react-devtools-shared/src/hooks/__tests__/updateMockSourceMaps.js [hunk 2]
```diff
@@ -323,7 +323,11 @@ async function bundle() {
     plugins: [
       rollupResolve(),
       commonjs(),
-      babel({presets: ['@babel/preset-react'], sourceMap: true}),
+      babel({
+        presets: ['@babel/preset-react'],
+        sourceMap: true,
+        babelHelpers: 'bundled',
```
Comment (ymqy) [inline packages/react-devtools-shared/src/hooks/__tests__/updateMockSourceMaps.js:329]: configure this option explicitly with its default value. https://github.com/rollup/plugins/tree/master/packages/babel#babelhelpers

File: packages/react-dom/src/server/ReactDOMFizzStaticNode.js [hunk 3]
```diff
@@ -10,7 +10,8 @@
 import type {ReactNodeList} from 'shared/ReactTypes';
 import type {BootstrapScriptDescriptor} from 'react-dom-bindings/src/server/ReactDOMServerFormatConfig';
 
-import {Writable, Readable} from 'stream';
+import type {Writable} from 'stream';
+import {Readable} from 'stream';
```
Comment (ymqy) [inline packages/react-dom/src/server/ReactDOMFizzStaticNode.js:14]: Fix build error
image

File: scripts/rollup/build.js [hunk 4]
```diff
@@ -185,7 +186,8 @@ function getRollupOutputOptions(
     format,
     globals,
     freeze: !isProduction,
-    interop: false,
+    interop: 'default',
```
Comment (ymqy) [inline scripts/rollup/build.js:189]: The interop configuration uses the default values to minimize the size increase of the production bundles and pass the ci, and to be honest I'm not very confident about this change.

File: scripts/rollup/build.js [hunk 5]
```diff
@@ -557,14 +562,15 @@ async function createBundle(bundle, bundleType) {
...
-  const pureExternalModules = Object.keys(importSideEffects).filter(
+  const moduleSideEffects = Object.keys(importSideEffects).filter(
     module => !importSideEffects[module]
   );
 
   const rollupConfig = {
     input: resolvedEntry,
     treeshake: {
-      pureExternalModules,
+      propertyReadSideEffects: false,
```
Comment (ymqy) [inline scripts/rollup/build.js:572]: I found some code added to the production bundles that reads variables but doesn't use them, so I set propertyReadSideEffects to false to eliminate them

File: scripts/rollup/build.js [hunk 6]
```diff
@@ -557,14 +562,15 @@ async function createBundle(bundle, bundleType) {
...
+      moduleSideEffects: id => !moduleSideEffects.includes(id),
```
Comment (ymqy) [inline scripts/rollup/build.js:573]: Personal opinion, the source code is best to ensure that the imported modules are no side effects, so we can set treeshake.moduleSideEffects to false, may be better, but I set false after the ci will not pass, now assign function type just to ci pass

**Comment by ymqy** [comment]
@eps1lon There are two other things that confuse me

1. It is not clear why some files were deleted

2. When building the output ReactDOMTesting-prod.js (fb_www_prod) file, errors are generated locally, but the ci passes3. 

image

https://app.circleci.com/pipelines/github/facebook/react/38285/workflows/1f6fe9f0-341a-456b-bec6-590112285dd3/jobs/627405/parallel-runs/31?filterBy=ALL

image

File: packages/react-client/src/forks/ReactFlightClientHostConfig.dom-node.js (hunk 1 above)
Comment (kassens) [reply packages/react-client/src/forks/ReactFlightClientHostConfig.dom-node.js:12]: These imports seem fragile to me and easy to mess. Why do we now need in some places the explicit `.js` suffix?

Does it have maybe something to do with this line below?

> moduleSideEffects: id => !moduleSideEffects.includes(id),

where the `id` only matches if the `.js` exists? (Just a guess, I don't really understand the config 😄 )

File: scripts/rollup/build.js (hunk 4 above)
Comment (kassens) [reply scripts/rollup/build.js:189]: From looking at the generated output files, I think this looks mostly right.

There's still a bit of extra code here (you can see the diffs in this comment https://github.com/facebook/react/pull/26078#issuecomment-1408052671 )

Example larger build where the change is visible: https://react-builds.vercel.app/commits/cc7dbd62a503bf502eb911b82a28926cf4d177e9/files/oss-stable/react-debug-tools/cjs/react-debug-tools.development.js?compare=4a4ef2706cfb51db96d48fe20dadcb4fb8e3cb17

**Comment by eps1lon** [comment]
> It is not clear why some files were deleted

No idea. I'd have to look through how the bundle configs are prepare to understand why some of them are missing now

> When building the output ReactDOMTesting-prod.js (fb_www_prod) file, errors are generated locally, but the ci passes3.

Maybe a caching issue (have you tested this on a new clone?). Maybe a different environment than CI?

**Comment by ymqy** [comment]
Try again from the beginning.

**Comment by markerikson** [comment]
I've been playing around with a Rollup v3 upgrade on my own machine, and I've run into the same issues:

- Need to change the `interop` flag
- Need to fix imports from `react-server` over into `react-server-dom-webpack/src` to deal with file extensions

But, there's another issue I'm seeing locally, and I just checked and it's happening in this PR, too.

A noticeable chunk of DOM attribute handling is getting stripped out of the `react-server-dom-webpack` bundles. Here's a screenshot of the diff from [one of the files in the size diffs earlier in this PR](https://react-builds.vercel.app/commits/6162fbede33758597d752598dc5dcf7a3054a4c6/files/oss-stable/react-server-dom-webpack/cjs/react-server-dom-webpack-server.node.development.js?compare=212b89fa25fd0a39473e845d1665d45f9da9dad3)

image

That seems like a pretty broken change, and I'm not immediately sure what's causing it.

**Comment by markerikson** [comment]
Per my above comment and https://github.com/facebook/react/issues/20186#issuecomment-1474400338 , I've got a Rollup v3 upgrade seemingly working locally as part of an attempt to generate sourcemaps.

I discussed with Dan, and on Monday I'm going to file 3 separate PRs for fixing bugs in the current Rollup 2.x build config, upgrading to Rollup 3.x, and adding sourcemaps.  That will supersede this PR, and I can work on getting those pushed through.

@ymqy : if you're around and available to finalize this PR in the next couple days, happy to use this one instead!  I just need this wrapped up quickly so I can use it as the basis for the sourcemap generation.

**Comment by ymqy** [comment]
@markerikson Thanks for the advice, I'll try to get as much done as possible.

**Comment by markerikson** [comment]
@ymqy Sure! I think the main thing that's missing in this PR atm is the bugfix for the `react-dom-webpack-server` output.  Not at my work computer atm, but the plugin I threw in was roughly this:

```ts
    {
        name: 'disable-treeshake',
        transform (code, id) {
            if (id.endsWith("DOMProperty.js") { 
                return {
                    code,
                    map: null,
                    moduleSideEffects: 'no-treeshake',
                };
            }
            return null;
        },
    }
```

**Comment by ymqy** [comment]
@markerikson I tried adding the disable-treeshake plugin and found that the react-dom-webpack-server build output still had problems. https://react-builds.vercel.app/commits/4cb72d38d6b52f98b0825ed93606be800497bc2a/files/oss-stable/react-server-dom-webpack/cjs/react-server-dom-webpack-server.node.development.js?compare=842bd787a55df9a93c4c6e0f9b003f100c564b5e

**Comment by markerikson** [comment]
@ymqy : no, the output looked good for me without `correctVarValueBeforeDeclaration`.  Busy for the next few hours, but let me grab the work laptop this afternoon and diff vs what you've got here.

**Comment by ymqy** [comment]
@markerikson OK, feel free to @ me if you have any questions.

> A noticeable chunk of DOM attribute handling is getting stripped out of the react-server-dom-webpack bundles.
I'm a bit confused, the problem is still there and the plugin doesn't solve the problem.

**Comment by markerikson** [comment]
@ymqy : ended up busy yesterday, but just fetched your branch and taking a look locally now. Will keep you posted on what I figure out!

**Comment by markerikson** [comment]
@ymqy : huh. If I check out your branch and build, I _do_ see the correct output (ie, the `properties` variable exists in the `react-dom-server-webpack` bundles, and I see `properties[name] = new PropertyInfoRecord()` being assigned).

Maybe the CI system is still reflecting an old artifact somehow?

The one other tweak I think _ought_ to be made to this PR is changing from `interop: 'default'` to `interop: 'esModule'`.  I compared the two, and it looks like `'default'` is adding a small helper util that isn't in the Rollup 2.x-based bundles, doesn't seem to be needed as far as I can see.

File: scripts/rollup/build.js (hunk 4 above)
Comment (markerikson) [reply scripts/rollup/build.js:192]: Pretty sure this should be `'esModule`' instead of `'default'`

**Comment by markerikson** [comment]
@ymqy : try rebasing this against `main` and making that `interop: 'esModule'` tweak, and see if the combination of those gets better results in CI?

**Comment by markerikson** [comment]
@ymqy : I filed my own Rollup v3 upgrade PR at #26442 , partly to see if the artifact diffs would come out better.

Good news is they did.  But, then my PR's CI jobs failed because I was using `interop: 'esModule'`, whereas yours uses `interop: 'default'`, and that broke some `react-noop-renderer` behavior.

I spent several more hours trying out several config tweaks to see if I could get Rollup to leave out the import interop helpers and still have `react-noop-renderer` run cleanly, and I got it working.  

I think that means #26442 is ready to be reviewed and merge at this point and would probably supersede this PR, but I specifically wanted to say that I based my changes on what you have here!  In particular, the trick of changing the `exports` map to point to `'./src/*.js` is a better solution than my first attempt of adding `.js` to a bunch of import statements.  So, thank _you_ for doing the initial work to get this upgrade rolling - it was very helpful!

**Comment by ymqy** [comment]
@markerikson Thank you very much for coming up with a better solution to omit the import interop helpers, and for learning from the big guy's solution.

I am also new to contributing to open source and feel honoured to be recognised.

At first I just wanted to upgrade Rollup, but I ran into some problems and needed to upgrade jest first, during which I raised a pr for jest to support the migration from jasmine to jest-circus, and only after this work was done could I start the rollup upgrade, which was additional to the initial work 😂

**Comment by markerikson** [comment]
@ymqy : hah, yeah, I totally understand that cascading series of "to fix tool X, I first have to update tool Y" problems :)

and yeah, I wanted to be clear that I'm not trying to come in and snatch away your chance to contribute here.  Even if _this_ PR doesn't get merged specifically, you did a lot of work to figure out what's needed for this upgrade, and I was able to build on top of _your_ work.  Really appreciate your efforts here!

**Comment by ymqy** [comment]
@markerikson I hadn't thought of it that way, and I'm glad you're building on my work. I Look forward to seeing you get the sourcemap generation working!

**Pull request closed without merging**